File: baekjoon/graph/21937/work.cpp
#include <iostream>
#include <vector>

typedef std::vector<int> t_vec;

const int   MAX = 100001;

t_vec map[MAX];
bool  visited[MAX];
int   vCnt, start, ans;

void    dfs(int cur)
{
    t_vec   &linked = map[cur];

    visited[cur] = true;
    for (const int &next: linked)
    {
        if (visited[next])
            continue ;

        ++ans;
        dfs(next);
    }
}

int main()
{
    std::cin.tie(0)->sync_with_stdio(0);

    int eCnt;

    std::cin >> vCnt >> eCnt;
    while (eCnt--)
    {
        int start, end;
        std::cin >> start >> end;
        map[end].push_back(start);
    }
    std::cin >> start;

    dfs(start);
    std::cout << ans << '\n';
    return 0;
}

File: baekjoon/graph/21937/work.py
import sys

sys.setrecursionlimit(10**5 + 1)

def dfs(cur: int) -> int:
    global visited

    ret = 0
    visited[cur] = True
    for next in edges[cur]:
        if visited[next]:
            continue
        ret += dfs(next) + 1

    return ret

if __name__ == '__main__':
    def readline(): return sys.stdin.readline()
    def splitIntoInt(): return map(int, readline().split())

    v_cnt, e_cnt = splitIntoInt()
    edges = [[] for _ in range(v_cnt + 1)]
    visited = [False for _ in range(v_cnt + 1)]
    while e_cnt:
        to, from_ = splitIntoInt()
        edges[from_].append(to)
        e_cnt -= 1
    start = int(readline())

    print(dfs(start))

